cmd/agent/app/integrations.go

```diff
@@ -0,0 +1,269 @@
+}
+
+func tuf(args []string) error {
+	if !allowRoot && !authorizedUser() {
+		return errors.New("Please use this tool as the agent-running user")
+	}
+
+	if flagNoColor {
+		color.NoColor = true
+	}
+
+	err := common.SetupConfig(confFilePath)
+	if err != nil {
+		fmt.Printf("Cannot setup config, exiting: %v\n", err)
+		return err
+	}
+
+	pipPath, err := getCommandPython()
+	if err != nil {
+		return err
+	}
+	tufPath, err := getTUFConfigFilePath()
+	if err != nil && !withoutTuf {
+		return err
+	}
+
+	cmd := args[0]
+	implicitFlags := args[1:]
+	args = append([]string{"-mpip"}, cmd)
+
+	// Add pip power-user flags
+	// cmd-flags go before the actual command
+	cmdFlags, err := tufCmd.Flags().GetStringSlice("cmd-flags")
+	if err == nil {
+		args = append(args, cmdFlags...)
+	}
+	args = append(args, implicitFlags...)
+
+	// Proxy support
+	proxies, err := config.GetProxies()
+	if err == nil && proxies != nil {
+		proxyFlags := fmt.Sprintf("--proxy=%s", proxies.HTTPS)
+		args = append(args, proxyFlags)
+	}
+
+	// idx-flags go after the command and implicit flags
+	idxFlags, err := tufCmd.Flags().GetStringSlice("idx-flags")
+	if err == nil {
+		args = append(args, idxFlags...)
+	}
+
+	tufCmd := exec.Command(pipPath, args...)
+
+	if !withoutTuf {
+		tufCmd.Env = append(os.Environ(),
+			fmt.Sprintf("TUF_CONFIG_FILE=%s", tufPath),
+		)
+	}
+
+	// forward the standard output to the Agent logger
+	stdout, err := tufCmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	go func() {
+		in := bufio.NewScanner(stdout)
+		for in.Scan() {
+			fmt.Println(in.Text())
+		}
+	}()
+
+	// forward the standard error to the Agent logger
+	stderr, err := tufCmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	go func() {
+		in := bufio.NewScanner(stderr)
+		for in.Scan() {
+			fmt.Println(color.RedString(in.Text()))
+		}
+	}()
+
+	err = tufCmd.Start()
+	if err != nil {
+		fmt.Printf("error running command: %v", err)
+		return err
+	}
+
+	err = tufCmd.Wait()
+	return err
+}
+
+func installTuf(cmd *cobra.Command, args []string) error {
+	constraintsPath, err := getConstraintsFilePath()
+	if err != nil {
+		return err
+	}
+
+	cachePath, err := getTUFPipCachePath()
+	if err != nil {
+		return err
+	}
+
+	tufArgs := []string{
+		"install",
+		"--cache-dir", cachePath,
+		"-c", constraintsPath,
+	}
+
+	tufArgs = append(tufArgs, args...)
+	if !withoutTuf {
+		tufArgs = append(tufArgs, "--index-url", tufPyPiServer)
+		tufArgs = append(tufArgs, "--extra-index-url", pyPiServer)
+		tufArgs = append(tufArgs, "--disable-pip-version-check")
+	}
+
+	return tuf(tufArgs)
+}
+
+func removeTuf(cmd *cobra.Command, args []string) error {
+	tufArgs := []string{
+		"uninstall",
+	}
+	tufArgs = append(tufArgs, args...)
+	tufArgs = append(tufArgs, "-y")
+
+	return tuf(tufArgs)
+}
+
+func searchTuf(cmd *cobra.Command, args []string) error {
+
+	tufArgs := []string{
+		"search",
+	}
+	tufArgs = append(tufArgs, args...)
+	if !withoutTuf {
+		tufArgs = append(tufArgs, "--index", tufPyPiServer)
```

Comment:
With pip+TUF, we don't need both the --index-url and --extra-index-url options :) I highly recommend removing them...

Also, I don't think `pip search` would work with a TUF repository, because it requires a web service to reply to POST queries to `/simple`, whereas the TUF repository is simply a static file server (indeed, we use an S3 bucket in this case).

One workaround is to download `/simple/index.html` with TUF, and then parse it. We can discuss this.

Reply:
It works with PyPi, so we'll keep it for now.